readthedocs/projects/tasks.py

```diff
@@ -593,7 +593,7 @@ def update_imported_docs(version_pk):
                 version_slug = version.slug
                 version_repo = project.vcs_repo(version_slug)
 
-                ret_dict['checkout'] = version_repo.checkout(version.identifier)
+                ret_dict['checkout'] = version_repo.checkout(version.identifier, env)
```

Comment:
Do we need the `setup_env` here, or could we just create a new `vcs_env` inside checkout?

Reply:
Seems we could pass it into the original call to `project.vcs_repo(version_slug)` in the line above, and stored on the repo object?

Reply:
My initial thought was to just use the `setup_env`. The difference between `setup_env` and `build_env` is just that one is executed on the host, and one in the vm. Separating this further might make for more places to check for env exception vs env failure vs env pass as well, so reusing `setup_env` ends up cleaner too.

Do you have a scenario that you think it might help to have a separate env?

Reply:
I think `setup_env` currently is just being used for VCS checkouts for the most part, so probably the proper place for it.

Reply:
The `setup_env` is not used for VCS checkouts since they are using their own `run`. In fact, the function/task didn't receive the `setup_env`, I had to added to it.

That could be a problem because that task (`update_imported_docs`) is called from a couple of places where a `setup_env` doesn't exist (e.g. from `core/views/hooks.py`). This could be one of the reason of why a `vcs_env` could be a better approach: we could create it inside the `Project.vcs_repo` instead of passing around through all the code or forcing other calls to create this environment just to log the commands.

Reply:
Something that we need to consider if we want to create a new `vcs_env` is that the envs are creating with this dependencies:

* `self.config` (from YAML)
* `env_vars`. From `get_env_vars`, which depends on
   * `self.version`
   * `self.project`
   * `self.config`
* `self.build`
* `record`

In case we want to create the environment inside the `Project.vcs_repo` will need to consider this dependencies.

Reply:
I took a look at this, and I found that the only missing piece here is the `build_pk` (originally passed from `core.utils.trigger_build` after creating an Build object).

### Solution 1

So, I'd say that this method `update_imported_docs` could receive the `version_pk` _and_ `build_pk` (but when called from `UpdateDocsTask` we will be repeating a couple of actions like parsing the YAML and more) and create our own `vcs_env` for this. If `build_pk` is `None` we won't `record` (`False`) the commads since we don't know where to record them. This will allow other uses of this function to keep working.

### Solution 2

Refactor all the things!

* move `update_imported_docs` as a method of `UpdateDocsTask`
* fix `test_celery.py:test_update_imported_docs`
* refactor/remove `core.view.hooks._build_url` function that uses the function `update_imported_docs`
* refactor/remove `github_build`, `gitlab_build` and `bitbucket_build` functions that use the `_build_url` but the docstrings say those functions are deprecated in favor of `readthedocs.restapi.views.integrations` classes (these classes use `core.views.hooks._build_version` which uses `trigger_build` which uses `UpdateDocsTask` --so we are safe)

I _think_ the last solution is a better plan, but I may missing a lot of things in the middle :)

Reply:
We found that there are users using the old webhook endpoints. Only projects with `Feature.ALLOW_DEPRECATED_WEBHOOK` can hit these ones. But anyway, we can't remove them yet.

So, the refactor would include changing calls to `projects.tasks.update_imported_docs`,

https://github.com/rtfd/readthedocs.org/blob/445683fc0c86b0de56779447a8eedf98024dba32/readthedocs/core/views/hooks.py#L125-L129

by `core.utils.trigger_build` which will end up calling `UpdateDocsTask` in a proper way and "everything should keep working"

Reply:
I think we need think more about _why_ we are calling into this function from other places, and build a supported API for that use case. 

I believe all the other calls are so that we can re-rsync the checkout on disk and the DB (to update versions on a push, for example). We should find a way to make that supported explicitly in the build tooling I think, so that we don't have to maintain a bunch of different logic. I _think_ the best approach here is to have some kind of mechanism on the doc builds to say "Do a resync of all the on-disk stuff, but don't actually build the docs", which is mostly what your **Solution 2** is talking about.